aioredis/parser.py

```diff
@@ -0,0 +1,161 @@
+        self.pos += 1
+        return val
+
+    def readline(self, size=None):
+        if size is not None:
+            if len(self.buf) < size + 2 + self.pos:
+                yield from self.waitsome(size + 2)
+            offset = self.pos + size
+            if self.buf[offset:offset+2] != b'\r\n':
+                raise self.error("Expected b'\r\n'")
+        else:
+            offset = self.buf.find(b'\r\n', self.pos)
+            while offset < 0:
+                yield from self.waitany()
+                offset = self.buf.find(b'\r\n', self.pos)
+        val = self.buf[self.pos:offset]
+        self.pos = 0
+        del self.buf[:offset + 2]
+        return val
+
+    def readint(self):
+        try:
+            return int((yield from self.readline()))
+        except ValueError as exc:
+            raise self.error(exc)
+
+    def error(self, msg):
+        self._err = self.protocolError(msg)
+        return self._err
+
+    def parse(self):
+        if self._err is not None:
+            raise self._err
+        ctl = yield from self.readone()
+        if ctl == b'+':
+            val = yield from self.readline()
+            if self.encoding is not None:
+                try:
+                    # FIXME: decoding is done when whole
+                    #       (multi-bulk) value is read
+                    return val.decode(self.encoding)
+                except UnicodeDecodeError:
+                    pass
+            return bytes(val)
+        elif ctl == b'-':
+            val = yield from self.readline()
+            return self.replyError(val.decode('utf-8'))
+        elif ctl == b':':
+            return (yield from self.readint())
+        elif ctl == b'$':
+            val = yield from self.readint()
+            if val == -1:
+                return None
+            val = yield from self.readline(val)
+            if self.encoding:
+                try:
+                    # FIXME: decoding is done when whole
+                    #       (multi-bulk) value is read
+                    return val.decode(self.encoding)
+                except UnicodeDecodeError:
+                    pass
+            return bytes(val)
+        elif ctl == b'*':
+            val = yield from self.readint()
+            if val == -1:
+                return None
+            arr = []
+            for _ in range(val):
+                arr.append((yield from self.parse()))
+            return arr
+        else:
+            raise self.error("Invalid first byte: {!r}".format(ctl))
+
+    def parse_one(self):
+        if self._gen is None:
+            self._gen = self.parse()
+        try:
+            self._gen.send(None)
+        except StopIteration as exc:
+            self._gen = None
+            return exc.value
+        except Exception:
+            self._gen = None
+            raise
+        else:
+            return False
+
+
+try:
+    import hiredis
+    Reader = hiredis.Reader
+except ImportError:
+    Reader = PyReader
```

Comment:
if only `install_requires=['hiredis; platform_implementation_name=cpython']` worked since python3.3...
there will be a hack (if statement) in setup.py to require hiredis only for `cpython` platform,
thus allowing pypy3.
I'm now trying to setup travis-ci with fresh pypy3...